packages/block/block.ts

```diff
@@ -0,0 +1,222 @@
+  mapGet$,
+  mapHas$,
+  mapSet$,
+} from './dom';
+import { renderToTemplate } from './template';
+import { AbstractBlock, EditType, Hole } from './types';
+import type { Edit, EditChild, Props, VElement } from './types';
+
+export const createBlock = (fn: (props?: Props) => VElement) => {
+  let lastProp: string;
+  let lastHole: Hole;
+  const vnode = fn(
+    new Proxy(
+      {},
+      {
+        // A universal getter will return a Hole instance if props[any] is accessed
+        // Allows code to identify holes in virtual nodes ("digs" them out)
+        get(_, prop: string): Hole {
+          if (prop === lastProp) return lastHole;
+          const hole = new Hole(prop);
+          lastProp = prop;
+          lastHole = hole;
+          return hole;
+        },
+      },
+    ),
+  );
+  const edits: Edit[] = [];
+
+  // Turns vnode into a string of HTML and creates an array of "edits"
+  // Edits are instructions for how to update the DOM given some props
+  const content = renderToTemplate(vnode, edits);
+  const template = document.createElement('template');
+  innerHTML$.call(template, content);
+  const root = template.content.firstChild as HTMLElement;
+
+  // Handles case for positioning text nodes. When text nodes are
+  // put into a template, they can be merged. For example,
+  // ["hello", "world"] becomes "helloworld" in the DOM.
+  // Inserts text nodes into the DOM at the correct position.
+  for (let i = 0, j = edits.length; i < j; ++i) {
+    const current = edits[i]!;
+    if (!current.inits.length) continue;
+    const el = getCurrentElement(current, root);
+    for (let k = 0, l = current.inits.length; k < l; ++k) {
+      const init = current.inits[k]!;
+      insertText(el, init.value, init.index);
+    }
+  }
+
+  return (props?: Props | null, key?: string) => {
+    return new Block(root, edits, props, key ?? props?.key);
+  };
+};
+
+export class Block extends AbstractBlock {
+  root: HTMLElement;
+  edits: Edit[];
+  constructor(
+    root: HTMLElement,
+    edits: Edit[],
+    props?: Props | null,
+    key?: string,
+  ) {
+    super();
+    this.root = root;
+    this.props = props;
+    this.edits = edits;
+    // Cache for getCurrentElement()
+    this.cache = new Map<number, HTMLElement>();
+    this.key = key;
+  }
+  mount(parent?: HTMLElement, refNode: Node | null = null): HTMLElement {
+    if (this.el) return this.el;
+    // cloneNode(true) uses less memory than recursively creating new nodes
+    const root = cloneNode$.call(this.root, true) as HTMLElement;
+
+    for (let i = 0, j = this.edits.length; i < j; ++i) {
+      const current = this.edits[i]!;
+      const el = getCurrentElement(current, root, this.cache, i);
+      for (let k = 0, l = current.edits.length; k < l; ++k) {
+        const edit = current.edits[k]!;
+        const hasHole = 'hole' in edit && edit.hole;
+        const value = hasHole ? this.props![edit.hole!] : undefined;
+
+        if (edit.type === EditType.Block) {
```

Comment:
It might be more efficient to use `if ... else if ... else if` or use a `switch` statement, because the JS engine will only have a single path to check